### PR TITLE
MM-20913 implement getbusy, setbusy, clearbusy

### PR DESCRIPTION
**Pull request by wiggin77**
#### Summary

This PR implements the following commands in mmctl:
- system getbusy
- system setbusy [seconds]
- system clearbusy

These APIs get/set/clear the busy state of the server, which disables/enables non-critical services when under high load.

Blocked by: https://github.com/mattermost/mattermost-server/pull/13986

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-20913

### CODE REVIEW

File: commands/system.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+	"github.com/spf13/cobra"
+
+	"github.com/mattermost/mmctl/client"
+	"github.com/mattermost/mmctl/printer"
+)
+
+var SystemCmd = &cobra.Command{
+	Use:   "system",
+	Short: "System management",
```
Comment (mgdelacroix) [inline commands/system.go]: Can we add a `Long` description here?

File: commands/system.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+var SystemGetBusyCmd = &cobra.Command{
+	Use:     "getbusy",
+	Short:   "Get the current busy state",
+	Long:    `Gets the server busy state (high load) and timestamp corresponding to when the server busy flag will be automatically cleared.`,
+	Example: `  system getbusy`,
```
Comment (mgdelacroix) [inline commands/system.go]:
```suggestion
	Example: `  system getbusy`,
	Args:    cobra.NoArgs,
```

File: commands/system.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	Example: `  system setbusy 3600`,
+	Args:    cobra.ExactArgs(1),
+	RunE:    withClient(setBusyCmdF),
+}
+
+var SystemClearBusyCmd = &cobra.Command{
+	Use:     "clearbusy",
+	Short:   "Clears the busy state",
+	Long:    `Clear the busy state, which re-enables non-critical services.`,
+	Example: `  system clearbusy`,
```
Comment (mgdelacroix) [inline commands/system.go]:
```suggestion
	Example: `  system clearbusy`,
	Args:    cobra.NoArgs,
```

File: commands/system.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	RunE:    withClient(clearBusyCmdF),
+}
+
+func init() {
+	SystemSetBusyCmd.Flags().Uint("seconds", 3600, "Number of seconds until server is automatically marked as not busy.")
```
Comment (mgdelacroix) [inline commands/system.go]: We could use here the shortflag format as well:

```suggestion
	SystemSetBusyCmd.Flags().UintP("seconds", "s", 3600, "Number of seconds until server is automatically marked as not busy.")
```

File: commands/system.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	SystemCmd.AddCommand(
+		SystemGetBusyCmd,
+		SystemSetBusyCmd,
+		SystemClearBusyCmd,
+	)
+	RootCmd.AddCommand(SystemCmd)
+}
+
+func getBusyCmdF(c client.Client, cmd *cobra.Command, args []string) error {
```
Comment (mgdelacroix) [inline commands/system.go]: I know this is inconsistently done in other commands, but if we are not using `args` we should use a blank assignment:

```suggestion
func getBusyCmdF(c client.Client, cmd *cobra.Command, _ []string) error {
```

File: commands/system.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	ok, response := c.SetServerBusy(int(seconds))
+	if response.Error != nil || !ok {
+		printer.PrintError(fmt.Sprintf("Unable to set busy state: %v", response.Error))
+		return response.Error
+	}
+	printer.Print("Busy state set")
+	return nil
+}
+
+func clearBusyCmdF(c client.Client, cmd *cobra.Command, args []string) error {
```
Comment (mgdelacroix) [inline commands/system.go]: Same:

```suggestion
func clearBusyCmdF(c client.Client, cmd *cobra.Command, _ []string) error {
```

File: commands/system.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	ok, response := c.ClearServerBusy()
+	if response.Error != nil || !ok {
+		printer.PrintError(fmt.Sprintf("Unable to clear busy state: %v", response.Error))
```
Comment (mgdelacroix) [inline commands/system.go]: There is the case of `response.Error == nil && ok == false` in which case we would be printing `nil` as part of the error message

File: commands/user_test.go [hunk 8]
```diff
@@ -591,7 +591,7 @@ func (s *MmctlUnitTestSuite) TestUserInviteCmd() {
 			Return(resultTeamModels[3], &model.Response{Error: nil}).
 			Times(1)
 
-		// Setup InvitUsersToTeam
+		// Setup InviteUsersToTeam
```
Comment (mgdelacroix) [inline commands/user_test.go]: Nice catch!

File: commands/system.go (hunk 4 above)
Comment (mgdelacroix) [reply commands/system.go]: We are setting here a flag for `seconds` but we are using a positional argument in the command itself. I think we should mark the flag as required, add the `Args: cobra.NoArgs` to the command and modify the logic to pick up seconds from the flag instead of `args[0]`

File: commands/system_test.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+			Times(1)
+
+		err := setBusyCmdF(s.client, &cobra.Command{}, []string{})
+		s.Require().Nil(err)
+		s.Require().Len(printer.GetLines(), 1)
+		s.Require().Equal(printer.GetLines()[0], map[string]string{"status": "ok"})
+		s.Require().Len(printer.GetErrorLines(), 0)
+	})
+
+	s.Run("SetBusy invalid seconds", func() {
```
Comment (larkox) [inline commands/system_test.go]: Probably corner case handled by ParseUInt, but what about a number > MaxInt?

File: commands/system_test.go [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+			Times(1)
+
+		err := clearBusyCmdF(s.client, &cobra.Command{}, []string{})
+		s.Require().Nil(err)
+		s.Require().Len(printer.GetLines(), 1)
+		s.Require().Equal(printer.GetLines()[0], map[string]string{"status": "ok"})
+		s.Require().Len(printer.GetErrorLines(), 0)
+	})
+
+	s.Run("ClearBusy with error", func() {
```
Comment (larkox) [inline commands/system_test.go]: On top of @mgdelacroix 's comment, you should add a test for error = nil, ok = false

File: commands/system_test.go [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+		s.client.
+			EXPECT().
+			GetServerBusy().
+			Return(nil, &model.Response{Error: &model.AppError{Id: "Mock Error"}}).
+			Times(1)
+
+		err := getBusyCmdF(s.client, &cobra.Command{}, []string{})
+		s.Require().NotNil(err)
+		s.Require().Len(printer.GetLines(), 0)
+		s.Require().Len(printer.GetErrorLines(), 1)
```
Comment (larkox) [inline commands/system_test.go]: Maybe I am nitpicking, but wouldn't it be nice to make sure the error you are receiving is the error you are expecting? Not sure if this will make the test too brittle, but may make sense.

File: commands/system_test.go [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+	s.Run("GetBusy with error", func() {
+		printer.Clean()
+		s.client.
+			EXPECT().
+			GetServerBusy().
+			Return(nil, &model.Response{Error: &model.AppError{Id: "Mock Error"}}).
+			Times(1)
+
+		err := getBusyCmdF(s.client, &cobra.Command{}, []string{})
+		s.Require().NotNil(err)
```
Comment (larkox) [inline commands/system_test.go]: I think we are preferring Require().Error(err) for these cases.

File: commands/system_test.go [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+		s.client.
+			EXPECT().
+			SetServerBusy(3600). // endpoint defaults to 3600 seconds
+			Return(true, &model.Response{Error: nil}).
+			Times(1)
+
+		err := setBusyCmdF(s.client, &cobra.Command{}, []string{})
+		s.Require().Nil(err)
+		s.Require().Len(printer.GetLines(), 1)
+		s.Require().Equal(printer.GetLines()[0], map[string]string{"status": "ok"})
```
Comment (larkox) [inline commands/system_test.go]: The order of the Equals is wrong. The definition is Equal(expected, actual). Therefore, in this example, it should be:
`s.Require().Equal(map[string]string{"status": "ok"}, printer.GetLines()[0])`

This may look like a nitpick, but it shows on the logs whenever the test fails.

File: commands/system_test.go [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+		const minutes = 15
+
+		s.client.
+			EXPECT().
+			SetServerBusy(minutes*60).
+			Return(true, &model.Response{Error: nil}).
+			Times(1)
+
+		err := setBusyCmdF(s.client, &cobra.Command{}, []string{strconv.Itoa(minutes * 60)})
+		s.Require().Nil(err)
```
Comment (larkox) [inline commands/system_test.go]: Similar to a previous comment, we are preferring here NoError over Nil

File: commands/system.go [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+	Example: `  system getbusy`,
+	RunE:    withClient(getBusyCmdF),
+}
+
+var SystemSetBusyCmd = &cobra.Command{
+	Use:     "setbusy [seconds]",
+	Short:   "Set the busy state to true",
+	Long:    `Set the busy state to true for the specified number of seconds, which disables non-critical services.`,
+	Example: `  system setbusy 3600`,
+	Args:    cobra.ExactArgs(1),
```
[the remark this answers is not in the document]
Comment (larkox) [reply commands/system.go]: After seeing @mgdelacroix message, if we use flags this would not be necessary.

File: commands/system.go (hunk 1 above)
Comment (wiggin77) [reply commands/system.go]: Fixed via https://github.com/mattermost/mmctl/pull/130/commits/62adb429e8db58d66c95d6dd13a8bd538fe46bd7

File: commands/system.go (hunk 2 above)
Comment (wiggin77) [reply commands/system.go]: Fixed via https://github.com/mattermost/mmctl/pull/130/commits/62adb429e8db58d66c95d6dd13a8bd538fe46bd7

File: commands/system.go (hunk 3 above)
Comment (wiggin77) [reply commands/system.go]: Fixed via https://github.com/mattermost/mmctl/pull/130/commits/62adb429e8db58d66c95d6dd13a8bd538fe46bd7

File: commands/system.go (hunk 4 above)
Comment (wiggin77) [reply commands/system.go]: Yes, that's much cleaner.  I haven't used Cobra much - getting a crash course from you. :-)

Fixed via https://github.com/mattermost/mmctl/pull/130/commits/ea2d91a248e9056585d42f34458be0a3b48cb4de

File: commands/system.go (hunk 5 above)
Comment (wiggin77) [reply commands/system.go]: Fixed via https://github.com/mattermost/mmctl/pull/130/commits/ea2d91a248e9056585d42f34458be0a3b48cb4de

File: commands/system.go (hunk 7 above)
Comment (wiggin77) [reply commands/system.go]: Ugh, right.  I don't remember why I made that API return a bool and an error.  Seems redundant now.

Comment (wiggin77) [reply commands/system.go]: Fixed via https://github.com/mattermost/mmctl/pull/130/commits/67d1566f460ead013fae90c82c7b37bb3e9f18a0    I double checked that the bool mirrors the error exactly (never returns false without error) so I'm just ignoring the bool.  Changing that API would be breaking so the behaviour won't change.  I don't know what I was thinking having that return a bool in the first place.

File: commands/system.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (wiggin77) [reply commands/system.go]: Thanks.  Using flag now https://github.com/mattermost/mmctl/pull/130/commits/ea2d91a248e9056585d42f34458be0a3b48cb4de

File: commands/system_test.go (hunk 10 above)
Comment (wiggin77) [reply commands/system_test.go]: See above.  The bool is redundant and I'm now ignoring it.

File: commands/system_test.go (hunk 13 above)
Comment (wiggin77) [reply commands/system_test.go]: Good catch.  I'll fix that for the whole file.

File: commands/system_test.go (hunk 9 above)
Comment (wiggin77) [reply commands/system_test.go]: Seconds is now a flag, and parsing is handled by Cobra.  It returns error.

File: commands/system_test.go (hunk 11 above)
Comment (wiggin77) [reply commands/system_test.go]: I did implement this, per our sidebar conversation. I ended up discarding it because it was brittle was what we gain.  Some cases are simple, some not so much.  I did however fix up the cases where I was returning an error and printing an error.

File: commands/system_test.go (hunk 12 above)
Comment (wiggin77) [reply commands/system_test.go]: Fixed

File: commands/system_test.go (hunk 13 above)
Comment (wiggin77) [reply commands/system_test.go]: Fixed

File: commands/system_test.go (hunk 14 above)
Comment (wiggin77) [reply commands/system_test.go]: Fixed

File: commands/system.go (hunk 7 above)
Comment (mgdelacroix) [reply commands/system.go]: I think I've seen the pattern someplace else in our API. I agree with ignoring the error being a good solution

File: commands/system.go [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+	sbs, response := c.GetServerBusy()
+	if response.Error != nil {
+		return fmt.Errorf("unable to get busy state: %w", response.Error)
+	}
+	printer.PrintT("busy:{{.Busy}} expires:{{.Expires}}", sbs)
+	return nil
+}
+
+func setBusyCmdF(c client.Client, cmd *cobra.Command, args []string) error {
```
Comment (mgdelacroix) [inline commands/system.go]: We should add `printer.SetSingle(true)` to this command too

File: commands/system.go [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+	_, response := c.SetServerBusy(int(seconds))
+	if response.Error != nil {
+		return fmt.Errorf("unable to set busy state: %w", response.Error)
+	}
+
+	printer.PrintT("Busy state set", map[string]string{"status": "ok"})
+	return nil
+}
+
+func clearBusyCmdF(c client.Client, cmd *cobra.Command, _ []string) error {
```
Comment (mgdelacroix) [inline commands/system.go]: Same here

File: commands/system.go (hunk 16 above)
Comment (wiggin77) [reply commands/system.go]: Fixed via https://github.com/mattermost/mmctl/pull/130/commits/cb820ee21914e2fa5792285f7538644a9f79f523

File: commands/system.go (hunk 17 above)
Comment (wiggin77) [reply commands/system.go]: Fixed via https://github.com/mattermost/mmctl/pull/130/commits/cb820ee21914e2fa5792285f7538644a9f79f523
